src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -657,9 +665,8 @@ void Assembler::emit_operand_helper(int reg_enc, int base_enc, int index_enc,
       }
     } else {
       // [base + disp]
-      assert(base_enc != rsp->encoding() LP64_ONLY(&& base_enc != r12->encoding()), "illegal addressing mode");
-      if (disp == 0 && no_relocation &&
-          base_enc != rbp->encoding() LP64_ONLY(&& base_enc != r13->encoding())) {
+      assert(((base_enc & 0x7) != 4), "illegal addressing mode");
+      if (disp == 0 && no_relocation &&  ((base_enc & 0x7) != 5)) {
```

Comment:
We loost information with this change. Can it be done as `is_r13_encoding(base_enc)` and `is_r12_enxoding(base_enc)`?

Reply:
Thanks for the comment.  The "& 0x7" style was suggested to me by @sviswa7 as a efficient way to check for r12, r20, r28 in the assert, and for r13, r21, r29 in the if statement.  I originally was comparing against each new APX register encoding.  The style in the PR is concise but it can be done either way.  What do you think?

Reply:
Got it - it is for few registers check now. What is common between rsp, r12, r20, r28 registers (except encoding)?
R12 is used for heap base in compressed oops  and RSP is RSP. What are r20 and r28? Why they can't be used in this addressing mode?

Please add comments for all lines where you replaced checks for `r*->encoding()` to say for which registers you do a check and why.

Reply:
The reason registers 12, 20, and 28 are asserted out of that else at line 668 is they are handled earlier in an else around line 646.

`    } else if ((base_enc & 0x7) == 4) {
      // [rsp + disp]
`

I added a comment to this effect and have also added comments in the 3 other places in the function where the replacement was done, indicating the registers involved.

Reply:
Good.